### ISSUE TITLE
Ion mobility from Mobilion mzML

### ISSUE THREAD

**Issue by chambm** [mzmine/mzmine#2169]
### Basic information

* My operating system and version: Windows 11
* My mzmine version:  4.3

### What happened

* Please describe what happened and how it differs from what you expected to happen.

I'm testing mzmine to check whether it can import converted files from Mobilion's MBI format. It seems these files aren't even recognized as ion mobility.

* Please include all parameter values for the mzmine modules you used. 

Defaults.

* Please upload screenshots of any errors **and** your log/mzmine.log file.

![image](https://github.com/user-attachments/assets/820ac5c4-4191-4b51-b5a7-389412bc0f82)
![image](https://github.com/user-attachments/assets/8bcb96e5-686f-4359-8979-cdf9ec8092d0)

Log:
```
2024-10-24 14:14:16 FINEST io.github.mzmine.gui.MZmineGUI activateSetOnDragDropped Importing 1 raw files via drag and drop: C:\dev\pwiz-mbi\20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.mzML
2024-10-24 14:14:16 FINEST io.github.mzmine.modules.io.import_rawdata_all.AllSpectralDataImportModule runModule File C:\dev\pwiz-mbi\20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.mzML type detected as MZML
2024-10-24 14:14:16 INFO   io.github.mzmine.taskcontrol.impl.WrappedTask run Starting processing of task Importing data
2024-10-24 14:14:16 INFO   io.github.mzmine.taskcontrol.impl.WrappedTask run Starting processing of task Importing raw data file: 20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.mzML
2024-10-24 14:14:16 FINEST io.github.mzmine.modules.io.import_rawdata_mzml.MSDKmzMLImportTask parseMzMl Began parsing file: C:\dev\pwiz-mbi\20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.mzML
2024-10-24 14:14:16 INFO   io.github.mzmine.modules.io.import_rawdata_mzml.msdk.data.MzMLParser processOpeningTag startTimeStamp value is: 2024-03-05T18:30:16Z
2024-10-24 14:14:17 FINEST io.github.mzmine.modules.io.import_rawdata_mzml.MSDKmzMLImportTask parseMzMlInternal Parsing Complete
2024-10-24 14:14:17 INFO   io.github.mzmine.modules.io.import_rawdata_mzml.MSDKmzMLImportTask importStreamOrFile Finished parsing C:\dev\pwiz-mbi\20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.mzML, parsed 5 scans
2024-10-24 14:14:17 FINEST io.github.mzmine.project.impl.MZmineProjectImpl addFile Adding a new file to the project: 20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.mzML
2024-10-24 14:14:17 INFO   io.github.mzmine.taskcontrol.impl.WrappedTask run Processing of task Importing 20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.mzML, total / parsed is 5 / 5 done, status FINISHED
2024-10-24 14:14:17 INFO   io.github.mzmine.taskcontrol.impl.WrappedTask run Processing of task Importing data done, status FINISHED
2024-10-24 14:14:18 FINEST io.github.mzmine.gui.mainwindow.MainWindowController handleShowRawDataOverview Activated Show raw data overview menu item
2024-10-24 14:14:18 FINEST io.github.mzmine.parameters.parametertypes.selectors.RawDataFilesSelection resetSelection Resetting file selection. Previously evaluated files: [20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.mzML]
2024-10-24 14:14:18 FINEST io.github.mzmine.main.MZmineCore runMZmineModule Module Raw data overview called at 2024-10-24T18:14:18.366856500Z
2024-10-24 14:14:18 FINEST io.github.mzmine.parameters.parametertypes.selectors.RawDataFilesSelection getMatchingRawDataFiles Setting file selection. Evaluated files: [20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.mzML]
2024-10-24 14:14:18 INFO   io.github.mzmine.taskcontrol.impl.WrappedTask run Starting processing of task Create raw data overview for 1 raw data files
2024-10-24 14:14:18 INFO   io.github.mzmine.taskcontrol.impl.WrappedTask run Processing of task Create raw data overview for 1 raw data files done, status FINISHED
2024-10-24 14:14:18 FINE   io.github.mzmine.modules.visualization.rawdataoverview.RawDataFileInfoPaneController populate Populating table for raw data file 20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.mzML
2024-10-24 14:14:18 FINE   io.github.mzmine.modules.visualization.rawdataoverview.RawDataOverviewWindowController addRawDataFileTab Added raw data file tab for 20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.mzML
2024-10-24 14:14:18 INFO   io.github.mzmine.modules.visualization.chromatogramandspectra.ChromatogramAndSpectraVisualizer lambda$setRawDataFiles$11 Change data files in visualizer
2024-10-24 14:14:18 FINEST io.github.mzmine.modules.visualization.chromatogramandspectra.ChromatogramAndSpectraVisualizer addRawDataFile Added raw data file 20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.mzML
2024-10-24 14:14:18 INFO   io.github.mzmine.taskcontrol.impl.WrappedTask run Starting processing of task Updating TIC visualizer of 20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.mzML
2024-10-24 14:14:18 INFO   io.github.mzmine.modules.visualization.chromatogram.TICDataSet run TIC data calculated for 20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.mzML
2024-10-24 14:14:18 INFO   io.github.mzmine.taskcontrol.impl.WrappedTask run Processing of task Updating TIC visualizer of 20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.mzML done, status FINISHED
```

* **Please upload a standalone mzmine project file (or a batch + raw files) that demonstrates the problem you are reporting.** You can use Google Drive, Dropbox, or a similar repository to upload the project.
* Issues opened without an example project are unlikely to be processed

Here's the subsetted mzML file shown above.
[20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.zip](https://github.com/user-attachments/files/17511176/20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.zip)

**Comment by SteffenHeu** [comment]
Hi,

i think this might be related to the merging of the ion mobility dimension. Is it possible to create a converted file including all the raw mobility spectra?

Looking at the file, the spectra look a bit odd. However, this might be related to the merging not beeing accounted for correctly.
![grafik](https://github.com/user-attachments/assets/5526883c-a2e9-4a71-8726-8a63a899e8f3)

**Comment by chambm** [comment]
Sorry I forgot to mention these are profile spectra without bounding zeros, so they don't plot that well. You want an unsubsetted file?

**Comment by SteffenHeu** [comment]
I think restricting the file to a few frames is not the issue, i think it's the merging of the mobility scans. Do you have a file without the merging? It can be a susbet of a few frames so the file isnt too large.

**Comment by chambm** [comment]
There are some new CV terms for Mobilion so maybe it's a CV update that's needed? Or does mzmine update that on the fly?

**Comment by chambm** [comment]
Here is the uncombined representation of the above file (same subset).
[20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.zip](https://github.com/user-attachments/files/17523891/20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.zip)

**Comment by SteffenHeu** [comment]
Hi,
thanks for the reply.

> We need to get the combined representation working: it's been the recommended one by the mzml group for a couple years.
> 
> It's quite simple and lossless. Instead of 2 arrays per spectrum and one spectrum per mobility bin, it's one spectrum per retention time (frame) and it has 3 arrays: m/z, intensity, and ion mobility. M/z values may be duplicated and mobility values are highly duplicated, but it's a vastly more efficient representation than one xml spectrum per mobility bin.

Yes we can work on it to support that. It would be good to know how all the values are organised in the arrays. I can see the mz and intensities have the same length, but mobility does have a different length. So how do they correspond to one another? How do we know how many of the mz/intensity values correspond to a single tof push so we can split them into spectra? What's the tag for the number of mobility spectra (if mobility values are duplicated)? Why are mobility values duplicated if a merged spectrum encompasses a single frame? That's why I'm asking for some documentation for the merged spectra. I looked through the mzml, but i can't see xml tags that would carry that information. Hope i did not miss anything.



> There are some new CV terms for Mobilion so maybe it's a CV update that's needed? Or does mzmine update that on the fly?

We would need to implement the new cv values manually to get it working properly. Do you have a list of new cv params already? I could find a tag for the ion mobility unit and type in the mzml

> Here is the uncombined representation of the above file (same subset). [20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.zip](https://github.com/user-attachments/files/17523891/20240305MJ_NISTmAb_2pmolCE25_03_SIFF25_0.zip)

It looks like the same combined file as before, the file sizes are equal. Can you please try to upload it again?

**Comment by chambm** [comment]
> Hi, thanks for the reply.
> 
> > We need to get the combined representation working: it's been the recommended one by the mzml group for a couple years.
> > It's quite simple and lossless. Instead of 2 arrays per spectrum and one spectrum per mobility bin, it's one spectrum per retention time (frame) and it has 3 arrays: m/z, intensity, and ion mobility. M/z values may be duplicated and mobility values are highly duplicated, but it's a vastly more efficient representation than one xml spectrum per mobility bin.
> 
> Yes we can work on it to support that. It would be good to know how all the values are organised in the arrays. I can see the mz and intensities have the same length, but mobility does have a different length. So how do they correspond to one another? 

The mobility array has the same length (arrayLength) as the defaultArrayLength (number of m/z and intensity points). If you see otherwise it's a bug.

> How do we know how many of the mz/intensity values correspond to a single tof push so we can split them into spectra? 

Each TOF push (mobility bin/scan) will have a different mobility. The arrays are usually sorted by mobility but you technically shouldn't count on that.

> What's the tag for the number of mobility spectra (if mobility values are duplicated)? Why are mobility values duplicated if a merged spectrum encompasses a single frame? That's why I'm asking for some documentation for the merged spectra. I looked through the mzml, but i can't see xml tags that would carry that information. Hope i did not miss anything.

There's no tag for that, you just have to parse the arrays.


> 
> > There are some new CV terms for Mobilion so maybe it's a CV update that's needed? Or does mzmine update that on the fly?
> 
> We would need to implement the new cv values manually to get it working properly. Do you have a list of new cv params already? I could find a tag for the ion mobility unit and type in the mzml
The new CV params are only for the file format type and nativeID format:
```
          <cvParam cvRef="MS" accession="MS:1001186" name="Mobilion MBI nativeID format" value=""/>
          <cvParam cvRef="MS" accession="MS:1001185" name="Mobilion MBI format" value=""/>
```

> It looks like the same combined file as before, the file sizes are equal. Can you please try to upload it again?

I edited the comment with the correct zip.

**Comment by SteffenHeu** [comment]
> The mobility array has the same length (arrayLength) as the defaultArrayLength (number of m/z and intensity points). If you see otherwise it's a bug.

Ok

> I edited the comment with the correct zip.

Thanks, with that file everything works as expected.


I looked at the merging option. I'm afraid with edge-cases like this: ![grafik](https://github.com/user-attachments/assets/f9c5c21d-2bc1-460b-8d08-86068854de64), which will cause deviation from the actual raw data structure when importing without special adaptations, it's not a quick and easy implementation. To achieve the same structural representation of the raw data in mzmine compared to importing from the raw data or from the un-optimised mzml, we would need to refactor large pieces of the import.

We can look at it in the future, but i cannot give a timeline for it.

If MSConvert supports the mbi files, we can call MSConvert from mzmine if the user has it installed and grab the mzml from stdout. I have created a PR for that in this repository, you can download the artifact and test it if you want. You can select your new MSConvert version in the preferences. Afterwards, it should automatically import.
PR is here: https://github.com/mzmine/mzmine/pull/2172

**Comment by chambm** [comment]
That deviation is only for Bruker DDA PASEF. Since you already support raw import of TDF, that shouldn't be as much of a priority. The uncombined representation for mzML produces hideously large files and is also much slower to convert (partly because I haven't bothered to optimize it for speed since it can't be optimized for file size). Thanks for adding the file extension support.

**Comment by SteffenHeu** [comment]
Ok it's good to know that the merged option is more optimised, that gives it some urgency. 

@chambm do you have a rough idea when msconvert will officially support the mbi files?

**Comment by SteffenHeu** [comment]
Hi @chambm,
i can't figure out where the type of the mobility data array is defined in the merged mbi file you provided. I mean which unit and analyzer it was. 
E.g. for the drift time in non-merged files i was able to use these accessions:
```
  public static final String cvMobilityDriftTime = "MS:1002476";
  public static final String cvMobilityDriftTimeUnit = "UO:0000028";
```

**Comment by chambm** [comment]
The third array type defines the metric. It should also define the units but I see that's missing (even though it can easily be inferred by the metric, e.g. drift time or 1/k0). I'll add the units though.

```
[Term]
id: MS:1003007
name: raw ion mobility array
def: "Array of raw ion mobility values (K or K0) based on ion separation in gaseous phase due to different ion mobilities under an electric field based on ion size, m/z and shape, corresponding to a spectrum of individual peaks encoded with an m/z array." [PSI:MS]
is_a: MS:1002893 ! ion mobility array
relationship: has_units UO:0000028 ! millisecond
relationship: has_units UO:0000010 ! second
relationship: has_value_type xsd:float ! The allowed value-type for this CV term

[Term]
id: MS:1003008
name: raw inverse reduced ion mobility array
def: "Array of raw ion mobility values based on ion separation in gaseous phase due to different ion mobilities under an electric field based on ion size, m/z and shape, normalized for the local conditions and reported in volt-second per square centimeter, corresponding to a spectrum of individual peaks encoded with an m/z array." [PSI:MS]
is_a: MS:1002893 ! ion mobility array
relationship: has_units MS:1002814 ! volt-second per square centimeter
relationship: has_value_type xsd:float ! The allowed value-type for this CV term
```

**Comment by chambm** [comment]
The third array will be fixed from:
`<cvParam cvRef="MS" accession="MS:1003007" name="raw ion mobility array" value=""/>`
to:
`<cvParam cvRef="MS" accession="MS:1003007" name="raw ion mobility array" value="" unitCvRef="UO" unitAccession="UO:0000028" unitName="millisecond"/>`

Do you need an updated file or can you do a quick replace all? :)

**Comment by SteffenHeu** [comment]
I can do a quick replace all, thanks for clarifying. 

When testing the results with other data formats, i noticed some issues that would prevent us from defaulting to the combined ims format. I opened an issue here: https://github.com/ProteoWizard/pwiz/issues/3209 It would be great if you could take a look and clarify what is going on.

**Comment by SteffenHeu** [comment]
will be included in #2200
